### PR TITLE
Use IKG for restarts, use IKG with ProtectedFileByteStore.

### PR DESCRIPTION
**Pull request by aam**
With super mixin issue with IKG kernel files resolved, we can come back to using IKG for app restarts.

Also this PR switches on use of ProtectedFileByteStore for IKG, which should improve performance of incremental builds. Use of byte store is only enabled for debug builds('build flx') and hot runs('run'), not for aot builds('build aot'). This is because aot builds require `platform.dill` linked into resultant dill file, which IKG doesn't seem to support(yet? @scheglov @sigmundch)

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_info.dart [hunk 1]
```diff
@@ -205,3 +205,8 @@ String getAssetBuildDirectory() {
 String getIosBuildDirectory() {
   return fs.path.join(getBuildDirectory(), 'ios');
 }
+
+/// Returns directory used by incremental compiler to store cached artifacts.
+String getIncrementalCompilerByteStoreDirectory() {
+  return fs.path.join(getBuildDirectory(), 'ikg_byte_store');
+}
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/build_info.dart]: seems to be missing newline at end of file.

File: packages/flutter_tools/lib/src/build_info.dart [hunk 2]
```diff
@@ -205,3 +205,8 @@ String getAssetBuildDirectory() {
 String getIosBuildDirectory() {
   return fs.path.join(getBuildDirectory(), 'ios');
 }
+
+/// Returns directory used by incremental compiler to store cached artifacts.
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/build_info.dart]: Is it really 'cached artifacts' or 'cached intermediate state' ?

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 3]
```diff
@@ -348,8 +348,19 @@ class HotRunner extends ResidentRunner {
     // TODO(aam): Add generator reset logic once we switch to using incremental
     // compiler for full application recompilation on restart.
     final bool updatedDevFS = await _updateDevFS(fullRestart: true);
-    if (!updatedDevFS)
+    if (!updatedDevFS) {
+      for (FlutterDevice device in flutterDevices) {
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/run_hot.dart]: The check for device.generator != null could be hoisted outside the loop
if (!updatedevFS) {
  if (device.generator != null) {
    for (FlutterDevice device in flutterDevices) {
      device.generator.reject();
    }
  }
  return new Ope.....
}

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 4]
```diff
@@ -348,8 +348,19 @@ class HotRunner extends ResidentRunner {
...
+        if (device.generator != null)
+          device.generator.reject();
+      }
       return new OperationResult(1, 'DevFS synchronization failed');
+    }
+    for (FlutterDevice device in flutterDevices) {
+      // VM must have accepted the kernel binary, there will be no reload
+      // report, so we let incremental compiler know that source code was accepted.
+      if (device.generator != null)
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/run_hot.dart]: ditto comment about hoisting the 'if' check outside the for loop.

**Review by a-siva — APPROVED** [review]

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 3 above)
Comment (aam) [reply packages/flutter_tools/lib/src/run_hot.dart]: `device` is the for-loop variable, so the has to be stay inside.

File: packages/flutter_tools/lib/src/build_info.dart (hunk 1 above)
Comment (aam) [reply packages/flutter_tools/lib/src/build_info.dart]: Done

File: packages/flutter_tools/lib/src/build_info.dart (hunk 2 above)
Comment (aam) [reply packages/flutter_tools/lib/src/build_info.dart]: Fixed the comment.

**Review by scheglov — APPROVED** [review]

**Pull request merged**